parcels/codegenerator.py

```diff
@@ -549,23 +549,28 @@ def generate(self, funcname, field_args, const_args, kernel_ast):
             args += [c.Pointer(c.Value("CField", "%s" % field))]
         for const, _ in const_args.items():
             args += [c.Value("float", const)]
-        fargs_str = ", ".join(['particles[p].time', 'sign * __dt'] + list(field_args.keys()) + list(const_args.keys()))
+        fargs_str = ", ".join(['particles[p].time', 'sign_dt * __dt'] + list(field_args.keys())
+                              + list(const_args.keys()))
         # Inner loop nest for forward runs
-        sign = c.Assign("sign", "dt > 0. ? 1. : -1.")
+        sign_dt = c.Assign("sign_dt", "dt > 0 ? 1 : -1")
+        sign_end_part = c.Assign("sign_end_part", "endtime - particles[p].time > 0 ? 1 : -1")
         dt_pos = c.Assign("__dt", "fmin(fabs(particles[p].dt), fabs(endtime - particles[p].time))")
         dt_0_break = c.If("particles[p].dt == 0", c.Statement("break"))
+        notstarted_continue = c.If("(sign_end_part != sign_dt) && (particles[p].dt != 0)",
+                                   c.Statement("continue"))
         body = [c.Assign("res", "%s(&(particles[p]), %s)" % (funcname, fargs_str))]
         body += [c.Assign("particles[p].state", "res")]  # Store return code on particle
-        body += [c.If("res == SUCCESS", c.Block([c.Statement("particles[p].time += sign * __dt"),
+        body += [c.If("res == SUCCESS", c.Block([c.Statement("particles[p].time += sign_dt * __dt"),
                                                  dt_pos, dt_0_break, c.Statement("continue")]))]
```

Comment:
Just to be sure: I think that we even did it together, but apart from checking that the test work, did you have a look at the generated c code and see that it is as expected? (it is so easy to do a typo in this generation script..)

Reply:
Yes, I checked the generated C code and all looks good